File: rust/2021/src/bin/day_09.rs
use aoc::geometry::{vecN, Vector};
use std::collections::{HashMap, HashSet};
use std::fs;

fn basin_size(grid: &HashMap<Vector<2>, u32>, low: Vector<2>) -> usize {
    let mut visited = HashSet::new();
    let mut to_visit = vec![low];
    while let Some(here) = to_visit.pop() {
        visited.insert(here);
        for adj in here.adjacents() {
            if let Some(&h) = grid.get(&adj) {
                if grid[&here] < h && h < 9 && !visited.contains(&adj) {
                    to_visit.push(adj);
                }
            }
        }
    }
    visited.len()
}

fn main() {
    let input = fs::read_to_string("../input/2021/input_09.txt").unwrap();
    let grid: HashMap<_, _> = input
        .lines()
        .enumerate()
        .flat_map(|(y, line)| {
            line.bytes()
                .enumerate()
                .map(move |(x, b)| (vecN!(x as i32, y as i32), (b - b'0') as u32))
        })
        .collect();
    let low_points: Vec<_> = grid
        .keys()
        .filter(|point| {
            point
                .adjacents()
                .all(|adj| grid[point] < *grid.get(&adj).unwrap_or(&9))
        })
        .collect();
    println!(
        "Part 1: {}",
        low_points.iter().map(|p| grid[p] + 1).sum::<u32>()
    );
    let mut basin_sizes: Vec<_> = low_points.iter().map(|&&p| basin_size(&grid, p)).collect();
    basin_sizes.sort();
    println!(
        "Part 2: {}",
        basin_sizes.iter().rev().take(3).product::<usize>()
    );
}

File: haskell/2021/day_09.hs
import qualified Data.Char as Char
import qualified Data.List as List
import qualified Data.Map as Map
import qualified Data.Set as Set

type Point = (Int, Int)

-- | Return a list of points adjacent to the given one.
findAdjacent :: Point -> Map.Map Point Int -> [Point]
findAdjacent (y, x) grid = filter (`Map.member` grid)
    [(y + dy, x + dx) | (dy, dx) <- [(0, 1), (1, 0), (0, -1), (-1, 0)]]

-- | Return a list of points that are lower than all their adjacent points.
lowPoints :: Map.Map Point Int -> [Point]
lowPoints grid = [point
    | (point, height) <- Map.assocs grid,
    and [height < grid Map.! adj | adj <- findAdjacent point grid]]

-- | Return a set of points that will flow downward to the given point.
findBasin :: Point -> Map.Map Point Int -> Set.Set Point
findBasin base grid =
    let height = grid Map.! base
        upper = [p | p <- findAdjacent base grid, grid Map.! p > height]
    in foldl Set.union (Set.singleton base) [findBasin p grid | p <- upper, grid Map.! p < 9]

main :: IO ()
main = do
    text <- readFile "../../input/2021/input_09.txt"
    let grid = Map.fromAscList $ do
        (y, line) <- zip [0..] $ lines text
        (x, c) <- zip [0..] line
        return ((y, x), Char.digitToInt c)
    putStrLn $ "Part 1: " ++ show (sum [grid Map.! p + 1 | p <- lowPoints grid])
    let sizes = Set.size <$> [findBasin p grid | p <- lowPoints grid]
    putStrLn $ "Part 2: " ++ show (product $ take 3 $ List.sortBy (flip compare) sizes)
